packages/esm-appointments-app/src/routes.json

```diff
@@ -39,6 +39,66 @@
       "name": "todays-appointments-dashboard",
       "slot": "todays-appointment-slot",
       "component": "homeAppointments"
+    },
+    {
+      "name": "expected-appointments-panel",
+      "slot": "scheduled-appointments-panels-slot",
+      "component": "appointmentsByStatus",
+      "meta": {
+        "title": "expected",
+        "status": "Scheduled",
+        "showForPastDate": false,
```

Comment:
I'm a bit of the mind that we should always show any tabs that have appointments in them.  Maybe we _also_ want to show some slots that have nothing in them too.  But if the backend has patients that have appointments for an earlier date that are still in the "scheduled" state, we kind of need to show them, otherwise there will be no way to clean those up manually, and there wouldn't be any visibility into that.  So we should think about this in terms of how this is set up.

Reply:
I'd probably prefer that we exposed this kind of configuration via the configuration system rather than `meta`. The configuration system is implementation-configurable, `meta` is not. There is a mechanism to allow a per-extension configuration schema, if that helps with this.

Reply:
@ibacher not sure I follow but I think you are saying that I can't do what I want users to be able to do, and therefore I need to rework.  I'd want someone to be able to do something like this in their own config. In this example they'd be removing one of the default extensions wired into this slot, and added two of their own.

```
"@openmrs/esm-appointments-app": {
    "extensionSlots": {
      "scheduled-appointments-panels-slot": {
        "remove": [
          "checked-in-appointments-panel"
        ],
        "add": [
          {
            "name": "my-custom-tab-using-existing-component",
            "component": "appointmentsByStatus",
            "meta": {
              "title": "My Custom Name",
              "status": "Completed",
              "showForPastDate": true,
              "showForToday": true,
              "showForFutureDate": true
            }
          },
          {
            "name": "my-custom-tab-using-my-new-component",
            "component": "myNewComponent"
          }
        ]
      }
    }
  },
```
I think you are saying this won't work, but, on top of that, how would you recommend doing this?  Thanks!

Reply:
The use-case for `meta` is somewhat different from this. In particular, its for passing information about how an extension should be rendered to the _slot_ rendering it, with the sort of canonical use-case for that being having a mechanism for saying "this extension should be rendered in one column" or something like that (which is a not-fully functional thing implemented in the [patient-chart](https://github.com/openmrs/openmrs-esm-patient-chart/blob/ee7b11ecc5ac709c74e3049a9a6b1f1759d9184e/packages/esm-patient-notes-app/src/routes.json#L12)). In other words, meta is for influencing the context in which the extension is rendered. Anything about how the extension itself functions should be part of the configuration system and fetched inside the extension via `useConfig()` in the more or less normal way.

`title` is a bit of an ambiguous property, but it's used in `meta` in cases where the slot itself will do something with that title. For example, the patient-chart dashboard mechanism uses the title as the name of the link in the left-nav, which is rendered independently of the extension itself.

Concretely, even if the above syntax worked, the limitation is that I have no way of changing `showForPastDate` (for example) in the `expected-appointments-panel`, only for newly added tabs. However, if `showForPastDate` is set via the configuration system I can set it for both newly added tabs and existing tabs declared in the routes.

Also note that the `add` property on config actually can't handle all of that. It's literally just a list of names of already-registered extensions. From the canonical docs on how this is supposed to work:

```json
{
  "@openmrs/esm-patient-chart-app": {
    "extensionSlots": {
      "patient-chart-dashboard-slot": {
        "add": [
          "nav-group#HIVAdultReturn",
          "nav-group#NCDOutpatient"
        ],
        "configure": {
          "nav-group#HIVAdultReturn": {
            "title": "HIV Adult Return",
            "slotName": "hiv-nav-group-slot"
          },
          "nav-group#NCDOutpatient": {
            "title": "NCD Outpatient",
            "slotName": "heart-nav-group-slot"
          }
        }
      }
    }
  }
}
```

There's no ability to influence `meta` here (because that's the role of the configuration system).

Reply:
Makes sense, thanks so much @ibacher.  Is there a pointer to an example of the pattern for the per-extension config schema?

Reply:
So, we currently don't have any production code I'm aware of using the feature, but there are some [feature tests](https://github.com/openmrs/openmrs-esm-core/blob/3fb6a3ac0ce437148f59da083ee266d34fae5e53/packages/framework/esm-config/src/module-config/module-config.test.ts#L1104), which are hopefully informative.

Basically, you'll just import `defineExtensionConfigSchema` from `@openmrs/esm-framework` and in `startupApp()` have something like:

```ts
defineExtensionConfigSchema("expected-appointments-panel", /* the config schema for this goes here */)
```

Then whenever someone provides a configuration for an `expected-appointments-panel`, they should be able to use that configuration schema like this:

```json
"@openmrs/esm-appointments-app": {
    "extensionSlots": {
      "scheduled-appointments-panels-slot": {
        "remove": [
          "checked-in-appointments-panel"
        ],
        "add": [
          "expected-appointments-panel#custom-tab"
         ],
         "configure": {
           "expected-appointments-panel#custom-tab": {
             "showForPastDate": true,
           }
         }
      }
    }
  },
```

Reply:
Thanks @ibacher ! I've got totally backlogged on stuff but hope to get back to this this week and will try this out!